### ISSUE TITLE
Not able to specify ::$allowed_parents, only ::$allowed_children on pages.

### ISSUE THREAD

**Issue by patricknelson** [silverstripe/silverstripe-cms#1167]
Not sure if this has affected anyone else, but I've noticed that there's no way to intuitively define which parents a child can have, if you need to. I've had a situation where I need all the flexibility/functionality of a page, but due to it's own functionality, should only be able to be added to a certain kind of parent. You can define the inverse relationship (that a certain parent can only have certain children) but you cannot restrict those same children from being added elsewhere without manually defining the relationship on ever single type of page in the CMS, which can be prohibitive and error prone from a development standpoint (e.g. if you forget to update one of the pages).

I found a pretty simple solution to this on my own by overriding the `->allowedChilren()` method and then using the following code:

```php
	public function allowedChildren() {
		// Default.
		$children = parent::allowedChildren();

		// Define allowed parents on a per page basis.
		$allowedParents = [];
		$pageTypes = SiteTree::page_type_classes();
		foreach($pageTypes as $pageType) {
			// Only worry about it if this page type has the private static ::$allowed_parents configured.
			$curParents = singleton($pageType)->stat("allowed_parents");
			if (!$curParents) continue;
			$allowedParents[$pageType] = $curParents;
		}

		// See if any one of the current allowed children exist in the array of pages above that restrict their parents.
		$restrictedChildren = array_intersect($children, array_keys($allowedParents));
		if (empty($restrictedChildren)) return $children; // ... nothing more to do.

		// Go through each of the children with restricted parents and remove it from the list of children if this class
		// DOES NOT appear in its list of allowed parents.
		$thisPage = get_class($this);
		foreach($restrictedChildren as $childPage) {
			if (!in_array($thisPage, $allowedParents[$childPage])) {
				$childIndex = array_search($childPage, $children);
				unset($children[$childIndex]);
			}
		}

		// Return processed array.
		return $children;
	}
```

Is this something that you guys would be interested in rolling out into the CMS somehow? I'm thinking that other people might find this useful as well, so maybe it could be implemented on `SiteTree` or as "yet another module".

**Comment by dhensby** [comment]
I do believe this is something that the CMS is missing and needs to be addressed.

If you could open a PR of this against 3 branch, it'd be a way to get good feedback on it.

**Comment by neilcreagh** [comment]
Any progress on Allowed Parents in Silverstripe? 

Using allowed_children is there a way to allow Page but not some of the pages that extend Page?

**Comment by dhensby** [comment]
@neilcreagh the progress is here: https://github.com/silverstripe/silverstripe-cms/pull/1171

**Comment by dhensby** [comment]
> Using allowed_children is there a way to allow Page but not some of the pages that extend Page?

yes - see docs: https://github.com/silverstripe/silverstripe-cms/blob/53daaf9d5557619ff502a0a41de3fea8f4c37391/code/model/SiteTree.php#L53-L67

**Comment by maxime-rainville** [comment]
https://github.com/silverstripe/silverstripe-cms/pull/1171 was almost over the line for SilverStripe 3.

I had a go to see if it could be rebase on 4, but it's probably easier to replicate the changes in a separate PR than to resolve the conflicts.

**Comment by bcairns** [comment]
Just wanted to report that the above code still works perfectly in SS4 (eg, dropped into Page) without needing a single change, allowing this in child classes:

```
    private static $allowed_parents = [
        MyAllowedParentClass::class
    ];
```

Also worth mentioning here, SS4 now has an updateAllowedChildren() extension point in SiteTree->allowedChildren(), so it would be quite easy to implement this functionality as a SiteTree extension.

**Comment by spekulatius** [comment]
Have you wrapped it into an extension @bcairns? I feel like there is a number of people interested in it.

**Comment by patricknelson** [comment]
Probably worth noting here (since I just discovered this today in unrelated work) that I had to also write code in SS 3.x to ensure that the `CMSMain` (via an extension in my case) actually respected my modifications to `->allowedChildren()` when attempting to add a new page from that main "Add New" button in the very top left section of the Site Tree. 

Here's the code I wrote just for that back in 2015 in case anyone else here finds it useful (since they may not realize this gap still exists even after updating that method) 😄 

```php
<?php
/**
 * This is needed because SilverStripe isn't consistently paying attention to the already provided method of
 * ->allowedChildren() to determine if you can even list a page type as being optional under the MAIN "Add page" button
 * menu (top left corner). This helps to ensure it's leveraging that one centralized point more canonically.
 *
 * @author	Patrick Nelson, patnelson@ebay.com
 * @since	2015-06-29
 */

class CMSMainExtension extends DataExtension {

	/**
	 * @param	array	$disallowedChildren
	 * @param	int		$parentID
	 */
	public function updateChildFilter(&$disallowedChildren, $parentID) {

		// Start with an array of all page types and exclude those not listed.
		$allTypes = array_values(SiteTree::page_type_classes());

		/** @var Page $page */
		$page = Page::get()->filter("ID", $parentID)->first();
		if (!$page) return;
		$excluded = array_diff($allTypes, $page->allowedChildren());

		// Combine this new list of exclusions with the original.
		$excluded = array_merge($excluded, $disallowedChildren);
		$disallowedChildren = array_unique($excluded);
	}


}
```

**Comment by spekulatius** [comment]
Hello @patricknelson 

thanks for sharing the code. I hope someone stumbles across it and finds it useful! Seems like this issue has been around for a while already. I haven't been able to spend more on resolving this really as I got already dragged into the next project :/

Cheers,
Peter
